Search scroll context not being closed

As of 2.0.0, scroll contexts aren't being closed (at least not quickly), leading to too many scrolls open.

Stacktrace (IP/hostnames purged):

```org.elasticsearch.transport.RemoteTransportException: [metadata-canary][10.1.2.3:9300][indices:data/read/search[phase/query]]
Caused by: org.elasticsearch.ElasticsearchException: Trying to create too many scroll contexts. Must be less than or equal to: [1000]. This limit can be set by changing the [search.max_open_scroll_context] setting.
at org.elasticsearch.search.SearchService.createAndPutContext(SearchService.java:542) ~[elasticsearch-7.5.0.jar:7.5.0]
at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:348) ~[elasticsearch-7.5.0.jar:7.5.0]
at org.elasticsearch.search.SearchService.lambda$executeQueryPhase$1(SearchService.java:340) ~[elasticsearch-7.5.0.jar:7.5.0]
at org.elasticsearch.action.ActionListener.lambda$map$2(ActionListener.java:146) ~[elasticsearch-7.5.0.jar:7.5.0]
at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:63) ~[elasticsearch-7.5.0.jar:7.5.0]
at org.elasticsearch.action.ActionRunnable.lambda$supply$0(ActionRunnable.java:58) ~[elasticsearch-7.5.0.jar:7.5.0]
at org.elasticsearch.action.ActionRunnable$2.doRun(ActionRunnable.java:73) ~[elasticsearch-7.5.0.jar:7.5.0]
at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) ~[elasticsearch-7.5.0.jar:7.5.0]
at org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:44) ~[elasticsearch-7.5.0.jar:7.5.0]
at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:773) ~[elasticsearch-7.5.0.jar:7.5.0]
at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) ~[elasticsearch-7.5.0.jar:7.5.0]
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
at java.lang.Thread.run(Thread.java:830) [?:?]
```

--- a/docs/content/_docs/config.md
+++ b/docs/content/_docs/config.md
@@ -574,6 +574,9 @@ templateName: <string> default = heroic-metadata
 
 # Automatically add index mappings and settings.
 configure: <bool> default = false
+
+# Number of results to return in each response when using Elasticsearch scrolling.
+scrollSize: <int> default = 1000
 ```
 
 ##### `<es_index_config>`

--- a/heroic-elasticsearch-utils/src/main/java/com/spotify/heroic/elasticsearch/AbstractElasticsearchMetadataBackend.java
+++ b/heroic-elasticsearch-utils/src/main/java/com/spotify/heroic/elasticsearch/AbstractElasticsearchMetadataBackend.java
@@ -74,21 +74,21 @@ public abstract class AbstractElasticsearchMetadataBackend extends AbstractElast
         final Function<SearchHit, T> converter
     ) {
 
-    final ScrollTransform<T> scrollTransform =
-        new ScrollTransform<>(
-            async,
-            limit,
-            converter,
-            scrollId -> {
-                // Function<> that returns a Supplier
-                return () -> {
-                    final ResolvableFuture<SearchResponse> future = async.future();
-                    connection.searchScroll(scrollId, SCROLL_TIME, bind(future));
-                    return future;
-                };
-            });
-            final ResolvableFuture<SearchResponse> future = async.future();
-            connection.execute(request, bind(future));
+        final ScrollTransform<T> scrollTransform =
+            new ScrollTransform<>(
+                async,
+                limit,
+                converter,
+                scrollId -> {
+                    // Function<> that returns a Supplier
+                    return () -> {
+                        final ResolvableFuture<SearchResponse> future = async.future();
+                        connection.searchScroll(scrollId, SCROLL_TIME, bind(future));
+                        return future;
+                    };
+                });
+        final ResolvableFuture<SearchResponse> future = async.future();
+        connection.execute(request, bind(future));
 
         return future.lazyTransform(scrollTransform);
     }
@@ -119,8 +119,7 @@ public abstract class AbstractElasticsearchMetadataBackend extends AbstractElast
         }
 
         @Override
-        public AsyncFuture<ScrollTransformResult<T>> transform(final SearchResponse response)
-            throws Exception {
+        public AsyncFuture<ScrollTransformResult<T>> transform(final SearchResponse response) {
             final SearchHit[] hits = response.getHits().getHits();
             final String scrollId = response.getScrollId();
 

--- a/heroic-elasticsearch-utils/src/main/java/com/spotify/heroic/elasticsearch/RestConnection.kt
+++ b/heroic-elasticsearch-utils/src/main/java/com/spotify/heroic/elasticsearch/RestConnection.kt
@@ -106,7 +106,8 @@ class RestConnection(
     override fun searchScroll(
         scrollId: String, timeout: TimeValue, listener: ActionListener<SearchResponse>
     ) {
-        client.scrollAsync(SearchScrollRequest(scrollId), options, listener)
+        val request = SearchScrollRequest(scrollId).scroll(timeout)
+        client.scrollAsync(request, options, listener)
     }
 
     override fun clearSearchScroll(scrollId: String): ActionFuture<ClearScrollResponse> {

--- a/heroic-elasticsearch-utils/src/main/java/com/spotify/heroic/elasticsearch/TransportConnection.java
+++ b/heroic-elasticsearch-utils/src/main/java/com/spotify/heroic/elasticsearch/TransportConnection.java
@@ -157,7 +157,7 @@ public class TransportConnection implements Connection<StringTerms> {
         @NotNull TimeValue timeout,
         @NotNull ActionListener<SearchResponse> listener
     ) {
-        SearchScrollRequest request = new SearchScrollRequest(scrollId);
+        SearchScrollRequest request = new SearchScrollRequest(scrollId).scroll(timeout);
         client.searchScroll(request, listener);
     }
 

--- a/heroic-test/src/main/java/com/spotify/heroic/test/AbstractMetadataBackendIT.java
+++ b/heroic-test/src/main/java/com/spotify/heroic/test/AbstractMetadataBackendIT.java
@@ -77,6 +77,8 @@ public abstract class AbstractMetadataBackendIT {
 
     private AsyncFramework async;
 
+    protected final int numSeries = 3;
+
     private final Series s1 = Series.of("s1", ImmutableMap.of("role", "foo"));
     private final Series s2 = Series.of("s2", ImmutableMap.of("role", "bar"));
     private final Series s3 = Series.of("s3", ImmutableMap.of("role", "baz"));

--- a/heroic-test/src/test/java/com/spotify/heroic/metadata/elasticsearch/AbstractMetadataBackendKVIT.java
+++ b/heroic-test/src/test/java/com/spotify/heroic/metadata/elasticsearch/AbstractMetadataBackendKVIT.java
@@ -60,6 +60,7 @@ public abstract class AbstractMetadataBackendKVIT extends AbstractMetadataBacken
                 .index(index)
                 .clientSetup(setupClient())
                 .build())
+            .scrollSize(numSeries / 2)
             .build();
     }
 }

--- a/metadata/elasticsearch/src/main/java/com/spotify/heroic/metadata/elasticsearch/ElasticsearchMetadataModule.java
+++ b/metadata/elasticsearch/src/main/java/com/spotify/heroic/metadata/elasticsearch/ElasticsearchMetadataModule.java
@@ -80,6 +80,8 @@ public final class ElasticsearchMetadataModule implements MetadataModule, Dynami
     private static final int DEFAULT_WRITE_CACHE_CONCURRENCY = 4;
     private static final long DEFAULT_WRITE_CACHE_MAX_SIZE = 30_000_000L;
 
+    private static final int DEFAULT_SCROLL_SIZE = 1000;
+
     private static final String DEFAULT_GROUP = "elasticsearch";
     private static final String DEFAULT_TEMPLATE_NAME = "heroic-metadata";
 
@@ -95,6 +97,7 @@ public final class ElasticsearchMetadataModule implements MetadataModule, Dynami
     private final String distributedCacheSrvRecord;
     private final int deleteParallelism;
     private final boolean configure;
+    private final int scrollSize;
 
     private static Supplier<BackendType> defaultSetup = MetadataBackendKV::backendType;
 
@@ -125,7 +128,8 @@ public final class ElasticsearchMetadataModule implements MetadataModule, Dynami
         @JsonProperty("deleteParallelism") Optional<Integer> deleteParallelism,
         @JsonProperty("templateName") Optional<String> templateName,
         @JsonProperty("backendType") Optional<String> backendType,
-        @JsonProperty("configure") Optional<Boolean> configure
+        @JsonProperty("configure") Optional<Boolean> configure,
+        @JsonProperty("scrollSize") Optional<Integer> scrollSize
     ) {
         this.id = id;
         this.groups = groups.orElseGet(Groups::empty).or(DEFAULT_GROUP);
@@ -138,6 +142,8 @@ public final class ElasticsearchMetadataModule implements MetadataModule, Dynami
             writeCacheDurationMinutes.orElse(DEFAULT_WRITE_CACHE_DURATION_MINUTES);
         this.writeCacheConcurrency = writeCacheConcurrency.orElse(DEFAULT_WRITE_CACHE_CONCURRENCY);
         this.writeCacheMaxSize = writeCacheMaxSize.orElse(DEFAULT_WRITE_CACHE_MAX_SIZE);
+
+        this.scrollSize = scrollSize.orElse(DEFAULT_SCROLL_SIZE);
 
         this.distributedCacheSrvRecord = distributedCacheSrvRecord.orElse("");
 
@@ -233,6 +239,13 @@ public final class ElasticsearchMetadataModule implements MetadataModule, Dynami
 
         @Provides
         @ElasticsearchScope
+        @Named("scrollSize")
+        public int scrollSize() {
+            return scrollSize;
+        }
+
+        @Provides
+        @ElasticsearchScope
         public RateLimitedCache<Pair<String, HashCode>> writeCache(HeroicReporter reporter) {
             final Cache<Pair<String, HashCode>, Boolean> cache = CacheBuilder
                 .newBuilder()
@@ -292,6 +305,7 @@ public final class ElasticsearchMetadataModule implements MetadataModule, Dynami
         private Optional<String> templateName = empty();
         private Optional<String> backendType = empty();
         private Optional<Boolean> configure = empty();
+        private Optional<Integer> scrollSize = empty();
 
         public Builder id(final String id) {
             checkNotNull(id, "id");
@@ -365,21 +379,27 @@ public final class ElasticsearchMetadataModule implements MetadataModule, Dynami
             return this;
         }
 
+        public Builder scrollSize(int scrollSize) {
+            this.scrollSize = of(scrollSize);
+            return this;
+        }
+
         public ElasticsearchMetadataModule build() {
             return new ElasticsearchMetadataModule(
-              id,
-              groups,
-              connection,
-              writesPerSecond,
-              rateLimitSlowStartSeconds,
-              writeCacheDurationMinutes,
-              writeCacheConcurrency,
-              writeCacheMaxSize,
-              distributedCacheSrvRecord,
-              deleteParallelism,
-              templateName,
-              backendType,
-              configure
+                id,
+                groups,
+                connection,
+                writesPerSecond,
+                rateLimitSlowStartSeconds,
+                writeCacheDurationMinutes,
+                writeCacheConcurrency,
+                writeCacheMaxSize,
+                distributedCacheSrvRecord,
+                deleteParallelism,
+                templateName,
+                backendType,
+                configure,
+                scrollSize
             );
         }
     }

--- a/metadata/elasticsearch/src/main/java/com/spotify/heroic/metadata/elasticsearch/MetadataBackendKV.java
+++ b/metadata/elasticsearch/src/main/java/com/spotify/heroic/metadata/elasticsearch/MetadataBackendKV.java
@@ -130,7 +130,6 @@ public class MetadataBackendKV extends AbstractElasticsearchMetadataBackend
     private static final String METADATA_TYPE = "metadata";
 
     private static final TimeValue SCROLL_TIME = TimeValue.timeValueMillis(5000);
-    private static final int SCROLL_SIZE = 1000;
 
     private final Groups groups;
     private final MetadataBackendReporter reporter;
@@ -139,6 +138,7 @@ public class MetadataBackendKV extends AbstractElasticsearchMetadataBackend
     private final RateLimitedCache<Pair<String, HashCode>> writeCache;
     private final boolean configure;
     private final int deleteParallelism;
+    private final int scrollSize;
 
     @Inject
     public MetadataBackendKV(
@@ -148,7 +148,8 @@ public class MetadataBackendKV extends AbstractElasticsearchMetadataBackend
         Managed<Connection> connection,
         RateLimitedCache<Pair<String, HashCode>> writeCache,
         @Named("configure") boolean configure,
-        @Named("deleteParallelism") int deleteParallelism
+        @Named("deleteParallelism") int deleteParallelism,
+        @Named("scrollSize") int scrollSize
     ) {
         super(async, METADATA_TYPE);
         this.groups = groups;
@@ -158,6 +159,7 @@ public class MetadataBackendKV extends AbstractElasticsearchMetadataBackend
         this.writeCache = writeCache;
         this.configure = configure;
         this.deleteParallelism = deleteParallelism;
+        this.scrollSize = scrollSize;
     }
 
     @Override
@@ -453,7 +455,7 @@ public class MetadataBackendKV extends AbstractElasticsearchMetadataBackend
             SearchRequest request = c.getIndex().search(METADATA_TYPE);
             request.scroll(SCROLL_TIME);
             request.source()
-                .size(limit.asMaxInteger(SCROLL_SIZE))
+                .size(limit.asMaxInteger(scrollSize))
                 .query(new BoolQueryBuilder().must(f));
 
             modifier.accept(request);
@@ -518,7 +520,7 @@ public class MetadataBackendKV extends AbstractElasticsearchMetadataBackend
         return observer -> connection.doto(c -> {
             SearchRequest request = c.getIndex().search(METADATA_TYPE).scroll(SCROLL_TIME);
             request.source()
-                .size(limit.asMaxInteger(SCROLL_SIZE))
+                .size(limit.asMaxInteger(scrollSize))
                 .query(new BoolQueryBuilder().must(filter));
 
             modifier.accept(request);

